File: test/parsergenerator/symboldiscriminator_test.scala
package pg.parsergenerator

import org.scalatest.FunSpec
import pg.parsergenerator.symboldiscriminator.genSymbolDiscriminator
import pg.{data, token}
import data.sample_language.{test_calc_grammar, test_empty_language, test_sample_grammar}
import token.{Token}
import symboldiscriminator.{isTerminalSymbol, isNonterminalSymbol}

class symboldiscriminator_test extends FunSpec {

  describe("SymbolDiscriminator test") {
    describe("test sample language") {
      val symbols = genSymbolDiscriminator(test_sample_grammar)
      it("S is Nonterminal") {
        assert(isNonterminalSymbol(symbols, "S"))
        assert(!isTerminalSymbol(symbols, "S"))
      }
      it("E is Nonterminal") {
        assert(isNonterminalSymbol(symbols, "E"))
        assert(!isTerminalSymbol(symbols, "E"))
      }
      it("LIST is Nonterminal") {
        assert(isNonterminalSymbol(symbols, "LIST"))
        assert(!isTerminalSymbol(symbols, "LIST"))
      }
      it("T is Nonterminal") {
        assert(isNonterminalSymbol(symbols, "T"))
        assert(!isTerminalSymbol(symbols, "T"))
      }
      it("HOGE is Nonterminal") {
        assert(isNonterminalSymbol(symbols, "HOGE"))
        assert(!isTerminalSymbol(symbols, "HOGE"))
      }
      it("SEMICOLON is Terminal") {
        assert(!isNonterminalSymbol(symbols, "SEMICOLON"))
        assert(isTerminalSymbol(symbols, "SEMICOLON"))
      }
      it("SEPARATE is Terminal") {
        assert(!isNonterminalSymbol(symbols, "SEPARATE"))
        assert(isTerminalSymbol(symbols, "SEPARATE"))
      }
      it("ATOM is Terminal") {
        assert(!isNonterminalSymbol(symbols, "ATOM"))
        assert(isTerminalSymbol(symbols, "ATOM"))
      }
      it("ID is Terminal") {
        assert(!isNonterminalSymbol(symbols, "ID"))
        assert(isTerminalSymbol(symbols, "ID"))
      }
      it("INVALID (not appear in grammar) is neither Nonterminal nor Terminal") {
        assert(!isNonterminalSymbol(symbols, "INVALID"))
        assert(!isTerminalSymbol(symbols, "INVALID"))
      }
      it("Check nonterminal symbols set") {
        val nt: Set[Token] = symbols.nonterminal_symbols
        for (symbol <- Array("S", "E", "LIST", "T", "HOGE")) {
          assert(nt.contains(symbol))
        }
        assert(nt.size == 5)
      }
      it("Check terminal symbols set") {
        val t: Set[Token] = symbols.terminal_symbols
        for (symbol <- Array("SEMICOLON", "SEPARATE", "ATOM", "ID")) {
          assert(t.contains(symbol))
        }
        assert(t.size == 4)
      }
    }
    describe("test sample language") {
      val symbols = genSymbolDiscriminator(test_calc_grammar)
      it("Check nonterminal symbols set 2") {
        val nt: Set[Token] = symbols.nonterminal_symbols
        for (symbol <- Array("EXP", "TERM", "ATOM")) {
          assert(nt.contains(symbol))
        }
        assert(nt.size == 3)
      }
      it("Check terminal symbols set 2") {
        val t: Set[Token] = symbols.terminal_symbols
        for (symbol <- Array("PLUS", "ASTERISK", "DIGITS", "LPAREN", "RPAREN")) {
          assert(t.contains(symbol))
        }
        assert(t.size == 5)
      }
    }
    describe("test empty language") {
      val symbols = genSymbolDiscriminator(test_empty_language.grammar)
      it("Check nonterminal symbols set 3") {
        val nt: Set[Token] = symbols.nonterminal_symbols
        assert(nt.contains("S"))
        assert(nt.size == 1)
      }
      it("Check terminal symbols set 3") {
        val t: Set[Token] = symbols.terminal_symbols
        assert(t.size == 0)
      }
    }
  }
}

File: test/yacc/symboldiscriminator_test.ml
open OUnit
open Utils
open Symboldiscriminator
open Sample_language
open Token
open Language

let test () =
  let symbols = genSymbolDiscriminator(test_sample_grammar) in
  "SymbolDiscriminator test" >::: [
    "test sample language" >::: [
      "S is Nonterminal" >:: begin fun () ->
        assert(isNonterminalSymbol(symbols, "S"));
        assert(not(isTerminalSymbol(symbols, "S")))
      end;
      "E is Nonterminal" >:: begin fun () ->
        assert(isNonterminalSymbol(symbols, "E"));
        assert(not(isTerminalSymbol(symbols, "E")))
      end;
      "LIST is Nonterminal" >:: begin fun () ->
        assert(isNonterminalSymbol(symbols, "LIST"));
        assert(not(isTerminalSymbol(symbols, "LIST")))
      end;
      "T is Nonterminal" >:: begin fun () ->
        assert(isNonterminalSymbol(symbols, "T"));
        assert(not(isTerminalSymbol(symbols, "T")))
      end;
      "HOGE is Nonterminal" >:: begin fun () ->
        assert(isNonterminalSymbol(symbols, "HOGE"));
        assert(not(isTerminalSymbol(symbols, "HOGE")))
      end;
      "SEMICOLON is Terminal" >:: begin fun () ->
        assert(not(isNonterminalSymbol(symbols, "SEMICOLON")));
        assert(isTerminalSymbol(symbols, "SEMICOLON"))
      end;
      "SEPARATE is Terminal" >:: begin fun () ->
        assert(not(isNonterminalSymbol(symbols, "SEPARATE")));
        assert(isTerminalSymbol(symbols, "SEPARATE"))
      end;
      "ATOM is Terminal" >:: begin fun () ->
        assert(not(isNonterminalSymbol(symbols, "ATOM")));
        assert(isTerminalSymbol(symbols, "ATOM"))
      end;
      "ID is Terminal" >:: begin fun () ->
        assert(not(isNonterminalSymbol(symbols, "ID")));
        assert(isTerminalSymbol(symbols, "ID"))
      end;
      "INVALID (not appear in grammar) is neither Nonterminal nor Terminal" >:: begin fun () ->
        assert(not(isNonterminalSymbol(symbols, "INVALID")));
        assert(not(isTerminalSymbol(symbols, "INVALID")))
      end;
      "Check nonterminal symbols set" >:: begin fun () ->
        let nt = symbols.nonterminal_symbols in
        ["S"; "E"; "LIST"; "T"; "HOGE"] |> List.iter (fun symbol ->
          assert(S.mem symbol nt)
        );
        assert_equal (S.cardinal nt) 5
      end;
      "Check terminal symbols set" >:: begin fun () ->
        let t = symbols.terminal_symbols in
        ["SEMICOLON"; "SEPARATE"; "ATOM"; "ID"] |> List.iter (fun symbol ->
          assert(S.mem symbol t)
        );
        assert_equal (S.cardinal t) 4
      end;
    ];
    "test sample language" >::: [
      "Check nonterminal symbols set 2" >:: begin fun () ->
        let symbols = genSymbolDiscriminator(test_calc_grammar) in
        let nt = symbols.nonterminal_symbols in
        ["EXP"; "TERM"; "ATOM"] |> List.iter (fun symbol ->
          assert(S.mem symbol nt)
        );
        assert_equal (S.cardinal nt) 3
      end;
      "Check terminal symbols set 2" >:: begin fun () ->
        let symbols = genSymbolDiscriminator(test_calc_grammar) in
        let t = symbols.terminal_symbols in
        ["PLUS"; "ASTERISK"; "DIGITS"; "LPAREN"; "RPAREN"] |> List.iter (fun symbol ->
          assert(S.mem symbol t)
        );
        assert_equal (S.cardinal t) 5
      end;
    ];
    "test empty language" >::: [
      "Check nonterminal symbols set 3" >:: begin fun () ->
        let {grammar} = test_empty_language in
        let symbols = genSymbolDiscriminator(grammar) in
        let nt = symbols.nonterminal_symbols in
        assert(S.mem "S" nt);
        assert_equal (S.cardinal nt) 1
      end;
      "Check terminal symbols set 3" >:: begin fun () ->
        let {grammar} = test_empty_language in
        let symbols = genSymbolDiscriminator(grammar) in
        let t = symbols.terminal_symbols in
        assert_equal (S.cardinal t) 0
      end;
    ];
  ]
